krun/scheduler.py

```diff
@@ -345,56 +345,72 @@ def run(self):
-            extra_env=self._make_pre_post_cmd_env()
-        )
+        # Run the pre-exec commands, the benchmark and the post-exec commands.
+        # These are wrapped in a try/except, so that the post-exec commands
+        # are always executed, even if an exception has occurred. We only
+        # reboot /after/ the post-exec commands have completed.
+        try:
+            # Run the user's pre-process-execution commands We can't put an ETA
+            # estimate in the environment for the pre-commands as we have not
+            # (and should not) load the results file into memory yet.
+            util.run_shell_cmd_list(
+                self.config.PRE_EXECUTION_CMDS,
+            )
+
+            # Results should never be in memory while a benchmark runs because
+            # their size (which gets large) increases over the course of the
+            # benchmark session. If results were loaded, then each benchmark
+            # would be subject to a different memory layout, which is not fair.
+            assert self.results is None
+
+            # We collect rough execution times separate from real results. The
+            # reason for this is that, even if a benchmark crashes it takes
+            # time and we need to account for this when making estimates. A
+            # crashing benchmark will give an empty list of iteration times,
+            # meaning we can't use 'raw_exec_result' below for estimates.
+            exec_start_time = time.time()
+            measurements, instr_data, flag = job.run(self.mailer, self.dry_run)
+            exec_end_time = time.time()
+
+            # Store results
+            self.results = Results(self.config, self.platform,
+                                   results_file=self.config.results_filename())
+            self.results.append_exec_measurements(job.key, measurements)
+            self.results.add_instr_data(job.key, instr_data)
+
+            if flag == "E":
+                self.results.error_flag = True
+
+            eta_info = exec_end_time - exec_start_time
+            if self.platform.hardware_reboots:
+                # Add time taken to wait for system to come up if we are in
+                # hardware-reboot mode.
+                eta_info += STARTUP_WAIT_SECONDS
+            self.add_eta_info(job.key, eta_info)
+
+            # We dump the json after each process exec so we can monitor the
+            # JSON file mid-run. It is overwritten each time.
+            self.manifest.update(flag)
+
+        except Exception:
+            error_info = sys.exc_info()
+            lines = ["Krun error during benchmark execution: %s\n" % str(error_info[1])]
+            for frame in traceback.format_tb(error_info[2]):
+                lines.append(frame)
+            msg = "".join(lines)
+            util.log_and_mail(self.mailer, debug,
+                              "Krun error during benchmark execution",
+                              msg, bypass_limiter=False)
+
```

Comment:
Surely the post-exec commands in the `finally` won't happen if we `sys.exit(1)` here?